### DISCUSSION TITLE
Type hints not working when querying if select is used directly in query

### DISCUSSION THREAD

**Discussion by bmakan** [sqlalchemy/sqlalchemy#10432] [Usage Questions] [answered]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

Type hints don't work consistently with pylance.

```python
class Location(Base):
    '''
    SQLA model.
    '''

    __tablename__ = 'locations'

    id: Mapped[int] = mapped_column(autoincrement=True, primary_key=True)
    name: Mapped[str] = mapped_column(sa.String(80), unique=True)


with scoped_session() as session:
    locations = session.scalars(sa.select(Location)).all()
    # (variable) locations: Sequence[Any]
    locations = session.execute(sa.select(Location)).all()
    # (variable) locations: Sequence[Row[Any]]
    locations = session.query(Location).all()
    # (variable) locations: List[Location]

    s = sa.select(Location)
    locations = session.scalars(s).all()
    # (variable) locations: Sequence[Location]
    locations = session.execute(s).all()
    # (variable) locations: Sequence[Row[Tuple[Location]]]
```

### To Reproduce

```python
import sqlalchemy as sa
import sqlalchemy.orm as saorm
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column

URI = f''

engine = sa.create_engine(URI)
session_factory = saorm.sessionmaker(bind=engine)
scoped_session = saorm.scoped_session(session_factory)

B = saorm.declarative_base()


class Location(B):
    '''
    SQLA model.

    Describes a physical location (a site).
    '''

    __tablename__ = 'locations'

    id: Mapped[int] = mapped_column(autoincrement=True, primary_key=True)
    name: Mapped[str] = mapped_column(sa.String(80), unique=True)


with scoped_session() as session:
    locations = session.scalars(sa.select(Location)).all()
    locations = session.execute(sa.select(Location)).all()
    locations = session.query(Location).all()

    s = sa.select(Location)
    locations = session.scalars(s).all()
    locations = session.execute(s).all()
```

### Error

```
# Copy the complete text of any errors received by the type checker(s).
``` 

### Versions

- OS: Ubuntu 20.04.6 LTS
- Python: 3.11.4
- SQLAlchemy: 2.0.21
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): Pylance language server 2023.10.10 (pyright 45942a15)

**Comment by bmakan** [comment]
When I added a `where` clause to the `select`, the typing was correctly identified.
But I also found instances where not even pulling the `select` statement into a variable helped.

**Comment by CaselIT** [comment]
Hi,

this seems a regression in pylance. It was definitely working at some point.

the fact that this does not work
```py
locations1 = session.scalars(sa.select(Location)).all()
```
but this does
```py
s = sa.select(Location)
locations2 = session.scalars(s).all()
```
indicates that the problem is not in sqlalchemy

You should probably report this to the pylance/pyright developers

**Reply by CaselIT** [reply, marked as answer]
using the previous version of the python extension seems to work, so this is a pyright/pylance issue

**Reply by bmakan** [reply]
You're right. They released a new version yesterday which broke it.

**Comment by zzzeek** [comment]
OK we can't just wave this off, with every pyright issue, we must always be concerned that pyright is intentionally making some strictness changes that work against us.

the test is not clear what im looking for,  but I guess the issue is that 

```
locations2 = session.execute(sa.select(Location)).all()
```

does not type correctly while 

```
    s = sa.select(Location)
    locations5 = session.execute(s).all()
```

does.   hard to argue that's not a pyright bug, but someone needs to report this.

**Reply by CaselIT** [reply]
pyright works correctly. Tested both an old version I had instealled locally and the latest version 1.1.330 and it types everything correctly.

The issue seems limited to pylance, that's only available via vscode (I think there is no cli for it)

**Reply by CaselIT** [reply]
I think the pylance issues go here https://github.com/microsoft/pylance-release/issues
